Drive train gyro

chloe6248 opened the pull request.

Commit:
Created angle control methods for DriveTrain (using the Navx)

-As well as P-Loop and other related constants

Commit:
Created commands for incrementing/decrementing P constants and assigned them to buttons

jameshochadel commented on src/org/usfirst/frc/team2399/robot/RobotMap.java:
```diff
@@ -63,9 +64,14 @@
+	public static double DRIVE_ANGLE_P = 0;
+	public static final double DRIVE_ANGLE_ERROR = 0;
+	public static final double DRIVE_MIXED_LINEAR = 0;
+	public static final double DRIVE_MIXED_ANGULAR = 0;
```
I think 73-74 should have some description of what they do. (edit: I wrote this before you explained what 'mixed' meant, but I think I stand by it. 'mixed' is a good, short word for what it intends to describe, but a little note somewhere would be helpful)

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java:
```diff
@@ -0,0 +1,52 @@
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
A brief Javadoc should be filled out before merge. I'm not totally sure what these classes are for; I'm guessing the PID-loop tuning?

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
-		currentTalon.setP(currentPConstant += RobotMap.DISTANCE_INCREMENT);
+	public void incrementRightDistancePConstant(){
+		double currentPConstant = rightFrontTalon.getP();
+		rightFrontTalon.setP(currentPConstant += RobotMap.DRIVE_DISTANCE_INCREMENT);
```
Separation of concerns = great! These are in the right place. May I suggest making this one line, though? You create the `currentPConstant` variable, then use it only once; you could simply call the function inline. Same goes for the next few methods as well.

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
+	
+	public double getCurrentAngle()
+	{
+		return Navx.getYaw();
```
I'd like a javadoc that describes the return value a bit more. E.g. if the method returns `90.0`, what does that actually mean? Is it relative to the initial position? Is it clockwise or CCW? (If this is actually an obvious thing that I'm just missing, let me know and I'll clear this comment)

jameshochadel replied:
Update: See http://www.kauailabs.com/public_files/navx-mxp/apidocs/java/ -> AHRS -> getYaw; return value is in range (-180.0, 180.0).

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
+	{
+		double newDesiredAngle;
+		
+		if(getCurrentAngle()-180 >= getDesiredAngle())
```
Note: After chatting with team, I'd like a group activity during which we simplify this method. It does work, but I am like 90% certain it can be done more simply.

wmarshall replied:
Ah. Shortest-path angle measurement. There's a way to do this with one conditional, but it's very unintuitive unless you've written this method for a bunch of different projects. Would be happy to help with this.

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
+	
+	public void moveToAngle()
+	{
+		double pOutput = calculateAngleError() * anglePConstant;
```
Hmm, I don't think this is how p-constants are supposed to work. Perhaps I am misunderstanding, or this is a different P from the PID loop, but that P should be fed into the Talons.

wmarshall replied:
This is actually okay. The reason that this is done here vs on the talons is that the control loop specified by this method deals with steering (left/right) while the talons are only aware of velocity (forward/back). Steering is an abstraction over velocity, so it makes sense to sit higher in the abstraction pyramid

jameshochadel replied:
Yep, I checked in with Chloe and that makes sense. (I really wish these comments had a 'resolve' button)

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
+	
+	public boolean isDriveAngleFinished()
+	{
+		if(Math.abs(calculateAngleError()) <= RobotMap.DRIVE_ANGLE_ERROR)
```
You can shorten this to just say `return Math.abs(calculateAngleError()) <= RobotMap.DRIVE_ANGLE_ERROR`

jameshochadel commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
-		return currentTalon.getP();
+	public void driveAtAngleToDistance()
+	{
+		double anglePOutput = calculateAngleError() * anglePConstant;
```
Same question as above r/e use of p-constant

wmarshall replied:
Same answer as above

jameshochadel commented on .classpath:
```diff
@@ -7,5 +7,6 @@
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/TalonSRXLibJava.jar"/>
+	<classpathentry kind="var" path="navx-mxp"/>
```
Just brainstorming, but it might not be a bad idea to have a "setting up the 2017-Season project" document somewhere; a README in the project repo would be perfect for this! (It'd be pretty simple; just point them to the WPI 2017 System Update stuff and the notes you made in the journals for installing this)

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java:
```diff
@@ -0,0 +1,52 @@
+	private boolean isIncrementing;
+	
+
+    public DriveAnglePAdjustment(boolean isIncrementing) {
```
Good param naming!

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java:
```diff
@@ -0,0 +1,54 @@
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Same as last one - brief Javadoc plz

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java:
```diff
@@ -0,0 +1,54 @@
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	if(isIncrementing == true)
```
Line 29 can be simpler! (I will leave the simplification as an exercise to the reader)

jameshochadel commented on src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java:
```diff
@@ -0,0 +1,52 @@
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	if(isIncrementing == true)
```
Line 29 can be simpler! (I will leave the simplification as an exercise to the reader)

Commit:
Code review changes implemented - pull request #5

wmarshall commented on src/org/usfirst/frc/team2399/robot/OI.java:
```diff
@@ -57,6 +65,10 @@ public static double getRightY() {
+	private static Button driveDistanceIncrementButt = new JoystickButton(shooterJoy, 5);
+	private static Button driveDistanceDecrementButt = new JoystickButton(shooterJoy, 3);
+	private static Button driveAngleIncrementButt = new JoystickButton(shooterJoy, 6);
+	private static Button driveAngleDecrementButt = new JoystickButton(shooterJoy, 4);
```
Why are we adjusting drive constants on the shooter joystick? Not a fail, just a question.

wmarshall commented on src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
+	{
+		double pOutput = calculateAngleError() * anglePConstant;
+		driveRightVelocity(-pOutput);
+		driveLeftVelocity(pOutput);
```
Why the +/-? what coordinate system are we using? Looks like CW = + , but this isn't documented anywhere.

wmarshall reviewed (changes requested):
This still needs some work. Not much, in that I think we should move it towards testing ASAP, to validate that our NavX works, but enough that it should not be merged in current state.

sc18 committed:
Print to SmartDashboard

sc18 committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into DriveTrainGyro

# Conflicts:
#	.classpath
#	src/org/usfirst/frc/team2399/robot/OI.java
#	src/org/usfirst/frc/team2399/robot/RobotMap.java
#	src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java

sc18 committed:
Began to Add PIDSubsystem and Integrate with Code

DriveTrain now extends PIDSubsystem, with the required methods added, as
well as some changes to the constructor.

sc18 committed:
Finished Implementing PIDController, added DriveAngle and removed DriveAnglePAdjustment

Created methods to run the PIDController and added a command to set the
angle. Created 3 buttons and test angles to test the controller as well
as enable control from the SmartDashboard. The new coded has been
commented.

sc18 committed:
Changed method in DriveAngle command from setTargetAngle to DriveAngle

sc18 committed:
Removed the driveAngle method in DriveTrain

Added methods to set Talon control modes. Can call PIDController methods
in DriveAngle on driveTrain as it is an exetention of PIDSubsystem

Commit:
Merge branch 'master' into DriveTrainGyro

# Conflicts:
#	.classpath
#	src/org/usfirst/frc/team2399/robot/OI.java
#	src/org/usfirst/frc/team2399/robot/RobotMap.java
#	src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java

sc18 committed:
Merge branch 'DriveTrainGyroPIDController' into DriveTrainGyro

Conflicts:
	.classpath
	src/org/usfirst/frc/team2399/robot/OI.java
	src/org/usfirst/frc/team2399/robot/RobotMap.java
	src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java

sc18 committed:
Αδδεδ ψομμεντ ον ΑδριωεΤραιν ανγλε προτατιον, βασεδ ον ψλοψκςισε βεινγ ριγητ.

sc18 committed:
Deleted DriveAngle created a method to set the P outputs for distance for the left and right Talon

Commit:
no message

Commit:
Revert "no message"

This reverts commit 27777d7d9531f6082930f05da498ee488c1994e5.

chloe6248 committed:
Updated .classpath

Commit:
Merge branch 'DriveTrainGyro' of https://github.com/frc2399/2017-Season into DriveTrainGyro

sc18 committed:
Created DriveAtAngleForDistance command

chloe6248 committed:
Updated .classpath

Commit:
Merge branch 'DriveTrainGyro' of https://github.com/frc2399/2017-Season into DriveTrainGyro

sc18 committed:
Finished coding DriveAtAngleForDistance command

Commit:
Merge branch 'DriveTrainGyro' of https://github.com/frc2399/2017-Season into DriveTrainGyro
